src/compiler/crystal/tools/init.cr

```diff
@@ -0,0 +1,181 @@
+        puts "invalid TYPE value: #{skeleton_type}"
+        puts opts
+        exit 1
+      end
+      skeleton_type
+    end
+
+    def self.fetch_required_parameter(opts, args, name)
+      if args.empty?
+        puts "#{name} is missing"
+        puts opts
+        exit 1
+      end
+      args.shift
+    end
+
+    class Config
+      property skeleton_type
+      property name
+      property dir
+      property author
+      property silent
+
+      def initialize
+        @skeleton_type = "none"
+        @name = "none"
+        @dir = "none"
+        @author = "none"
+        @silent = false
+      end
+
+      def initialize(@skeleton_type, @name, @dir, @author, @silent = false)
+      end
+    end
+
+    class InitProject
+      getter config
+
+      def initialize(@config)
+      end
+
+      def run
+        views.each do |view|
+          view.new(config).render
+        end
+      end
+
+      def views
+        [
+         GitignoreView,
+         LicenseView,
+         ReadmeView,
+         TravisView,
+         ProjectileView,
+
+         SrcExampleView,
+         SrcVersionView,
+
+         SpecHelperView,
+         SpecExampleView,
+
+         GitInitView,
+        ]
+      end
+    end
+
+    abstract class View
+      getter config
+
+      def initialize(@config)
+      end
+
+      def render
+        Dir.mkdir_p(File.dirname(full_path))
+        File.write(full_path, to_s)
+        puts log_message unless config.silent
+      end
+
+      def log_message
+        "      #{"create".colorize(:light_green)}  #{full_path}" 
+      end
+
+      def module_name
+        config.name.camelcase
+      end
+
+      abstract def full_path
+    end
+
+    class GitInitView < View
+      getter config
+
+      def initialize(@config)
+      end
+
+      def render
+        return unless system(WHICH_GIT_COMMAND)
+        return command if config.silent
+        puts command
+      end
+
+      def full_path
+        "#{config.dir}/.git"
+      end
+
+      private def command
+        `git init #{config.dir}`
+      end
+    end
+
+    TEMPLATE_DIR = "#{__DIR__}/init/template"
+
+    macro template(name, template_path, full_path)
+      class {{name.id}} < View
+        ecr_file "{{TEMPLATE_DIR.id}}/{{template_path.id}}"
+        def full_path
+          "#{config.dir}/#{{{full_path}}}"
+        end
+      end
+    end
+
+    template GitignoreView, "gitignore.ecr", ".gitignore"
+    template LicenseView, "license.ecr", "LICENSE"
+    template ReadmeView, "readme.md.ecr", "README.md"
+    template TravisView, "travis.yml.ecr", ".travis.yml"
+    template ProjectileView, "projectfile.ecr", "Projectfile"
+
+    template SrcExampleView, "example.cr.ecr", "src/#{config.name}.cr"
+    template SrcVersionView, "version.cr.ecr", "src/#{config.name}/version.cr"
+
+    template SpecHelperView, "spec_helper.cr.ecr", "spec/spec_helper.cr"
+    template SpecExampleView, "example_spec.cr.ecr", "spec/#{config.name}_spec.cr"
```

Comment:
You could use this:

```crystal
abstract class Foo
end

class Bar < Foo
end

class Baz < Foo
end

macro def foo_subclasses : Array(Foo.class)
  {{ Foo.subclasses }}
end

p foo_subclasses #=> [Bar, Baz]
```

It still can't be written generically, as in `Foo.subclasses` because of some limitations/bugs.